functorch/_src/aot_autograd.py

```diff
@@ -1066,25 +1074,145 @@ def format_guard_bug_msg(aot_config, expected):
     )
 
 
-# Wraps aot_dispatch_deduplicated_autograd, ensuring that duplicate arguments
-# are dropped from the inner compilation function.
+# MOTIVATION:
 #
-# In Haskell types, suppose you have:
+# When tracing functions for future execution, one must be careful not to pass
+# in the same input tensor multiple times (e.g., f(x, x), as this can result
+# in graphs that are ONLY valid if you later pass a new tensor in exactly the
+# same way (e.g., f(y, y)).  (NB: we really mean duplicate; two distinct
+# tensors that alias each other is a different situation that is covered by
+# aot_dispatch_deduplicated_autograd). Here are two examples:
 #
-#   add_dupe_args :: DedupedArgs -> Args
-#   remove_dupe_args :: Args -> DedupedArgs
+# (1) Suppose you have a function:
 #
-#   aot_dispatch_deduplicated_autograd
-#       :: (DedupedArgs -> R) -> DedupedArgs -> AOTConfig -> (DedupedArgs -> R)
-#   aot_dispatch_autograd
-#       :: (Args -> R) -> Args -> AOTConfig -> (Args -> R)
+#   def f(x, y):
+#       return x + y
 #
-# Then the code below can be written in point-free style as:
+# If you make_fx(f)(x, x), you will trace out:
 #
-#   aot_dispatch_deduplicate_autograd f a c =
-#       aot_dispatch_autograd (f . add_dupe_args) (remove_dupe_args a) c . remove_dupe_args
+#   def f(x, y):
+#       return y + y
 #
-def aot_dispatch_autograd(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig):
+# Oops!
+#
+# (2) For most tensors x and y, you can compute f's gradient with respect to
+# these to inputs by saying torch.autograd.grad(f(x, y), (x, y)).  However,
+# if x is y, you will trace out a program that gets incorrect gradients:
+#
+#   >>> x = torch.randn(1, requires_grad=True)
+#   >>> torch.autograd.grad(x + x, (x, x))
+#   (tensor([2.]), tensor([2.]))
+#
+# In other words, the gradient is double-counted.  Deduplicating the arguments
+# gives you an appropriate gradient:
+#
+#   >>> y = torch.randn(1, requires_grad=True)
+#   >>> torch.autograd.grad(x + y, (x, y))
+#   (tensor([1.]), tensor([1.]))
+#
+# HOW TO DEDUPLICATE:
+#
+# There are a few strategies, in order of preference:
+#
+# 1. For every duplicate argument to the function, detach it into
+#    a separate leaf tensor, so that it is no longer duplicated.
+#
+#       PRO: The resulting compiled graph works for any configuration
+#       of duplicated arguments.
+#
+#       CON: It does not (naively) work if you mutate the metadata of inputs:
+#
+#           def f(x, y):
+#               x.transpose_(0, 1)
+#               y.transpose_(0, 2)
+#
+#           x = torch.randn(2, 3, 4)
+#           f(x, x)
+#
+#       The ordering of the transposes inside f dictates whether or not
+#       you get [4, 2, 3] or [3, 4, 2].  This means that you cannot precompute
```

Comment:
In theory, we could probably track the ordering that metadata mutation happened to inputs inside of `run_functionalized_fw_and_collect_metadata`, and replay them in that order? But that feels even more complicated than the 2-step thing that you're doing here.

Reply:
I think this is the correct long term solution, it allows us to avoid dupe specializing entirely.